[API Feature]: Multiblock API

Lanse505 opened the pull request:
This is an attempt at an Multiblock API. 
Inspired by Blusunrize's IE Multiblock API but built for Titanium with 1.15 in-mind.

Lanse505 committed:
Deprecated old IMultiblock stuff

Lanse505 committed:
Started work on new IShapedMultiblock API

Lanse505 committed:
Gone through a few iterations of ideas, all of them are doodoo

Lanse505 committed:
Reduced the Multiblock API down to it's core(ish)

Lanse505 committed:
Merge remote-tracking branch 'origin/1.15' into api/feature/multiblock

Lanse505 committed:
Merge 1.15 and push some misc changes

Lanse505 committed:
Tiny push (removes junk)

Lanse505 committed:
forgot ;

Lanse505 committed:
Started work on Multiblock Code

Lanse505 committed:
Continued work on Multiblock Code #SlowlyGettingThere

Lanse505 committed:
some small updates

SkySom commented on src/main/java/com/hrznstudio/titanium/api/multiblock/IMultiblock.java:
```diff
@@ -16,6 +16,7 @@
 
 import java.util.function.Predicate;
 
+@Deprecated
```
It's a new MC version, just kill this.

SkySom commented on src/main/java/com/hrznstudio/titanium/api/multiblock/IMultiblockHandler.java:
```diff
@@ -0,0 +1,63 @@
+public interface IMultiblockHandler {
+
+    /**
+     * @param state // Trigger BlockState
```
Why the additional //

SkySom commented on src/main/java/com/hrznstudio/titanium/api/multiblock/MultiblockTemplate.java:
```diff
@@ -0,0 +1,156 @@
+import java.util.List;
+
+/**
+ * Big shout-outs and credits to Immersive Engineering for providing most of the inspiration and example implementation code for us to figure out this implementation.
```
Are you respecting their license if you're pulling that much code that you have to shout out?

SkySom commented on src/main/java/com/hrznstudio/titanium/api/multiblock/MultiblockTemplate.java:
```diff
@@ -0,0 +1,156 @@
+            try {
+                template = StaticTemplateUtil.loadStaticTemplate(id);
+            } catch (IOException e) {
+                throw new RuntimeException(e);
```
So you need to crash the game here?

SkySom commented on src/main/java/com/hrznstudio/titanium/api/multiblock/TitaniumMultiblockTemplate.java:
```diff
@@ -0,0 +1,63 @@
+    }
+
+    private boolean isTileEntityMultiblock(TileEntity tileEntity) {
+        return tileEntity instanceof MachineControllerTile || tileEntity instanceof MachineFillerTile;
```
Why is this hardcoded?

SkySom commented on src/main/java/com/hrznstudio/titanium/api/multiblock/MachineMultiblock.java:
```diff
@@ -17,6 +17,7 @@
 import javax.annotation.Nullable;
 import java.util.function.Predicate;
 
+@Deprecated
```
Just kill it.

SkySom commented on src/main/java/com/hrznstudio/titanium/_impl/test/multiblock/MultiblockTestHandler.java:
```diff
@@ -0,0 +1,57 @@
+
+import javax.annotation.Nullable;
+
+public class MultiblockTestHandler implements IMultiblockHandler {
```
This seems like a broken test?

SkySom commented on src/main/java/com/hrznstudio/titanium/api/multiblock/matcher/ResultHandler.java:
```diff
@@ -0,0 +1,55 @@
+    final Event.Result result;
+    public static final ResultHandler DEFAULT = new ResultHandler(0, Event.Result.DEFAULT);
+
+    public ResultHandler(int x, Event.Result result) {
```
What is x?

SkySom commented on src/main/java/com/hrznstudio/titanium/api/multiblock/matcher/ResultHandler.java:
```diff
@@ -0,0 +1,55 @@
+        } else if(Math.abs(a.x) < Math.abs(b.x)) {
+            return b;
+        } else {
+            Titanium.LOGGER.info("Can't combine same valued results");
```
Why can't you?

SkySom commented on src/main/java/com/hrznstudio/titanium/multiblock/block/MachineControllerBlock.java:
```diff
@@ -0,0 +1,16 @@
+
+    @Override
+    public IFactory<T> getTileEntityFactory() {
+        return null;
```
This really should be nonnull.

SkySom commented on src/main/java/com/hrznstudio/titanium/multiblock/block/MachineFillerBlock.java:
```diff
@@ -0,0 +1,30 @@
+    }
+
+    @Override
+    public IFactory getTileEntityFactory() {
```
Should be nonnull

SkySom commented on src/main/java/com/hrznstudio/titanium/multiblock/tile/MachineControllerTile.java:
```diff
@@ -0,0 +1,80 @@
+        this.originalState = originalState;
+    }
+
+    public void addChild(MachineFillerTile child) {
```
Why is this hardcoded?

SkySom commented on src/main/java/com/hrznstudio/titanium/multiblock/tile/MachineControllerTile.java:
```diff
@@ -0,0 +1,80 @@
+
+    @Override
+    public ActionResultType onActivated(PlayerEntity player, Hand hand, Direction facing, double hitX, double hitY, double hitZ) {
+        if (super.onActivated(player, hand, facing, hitX, hitY, hitZ) == ActionResultType.FAIL) {
```
What's the chances it's just ActionResultType.PASS

SkySom commented on src/main/java/com/hrznstudio/titanium/multiblock/tile/MachineControllerTile.java:
```diff
@@ -0,0 +1,80 @@
+
+    @Nonnull
+    @Override
+    public <U> LazyOptional<U> getCapability(@Nonnull Capability<U> cap, @Nullable Direction side) {
```
Why overriding just to call super?

SkySom commented on src/main/java/com/hrznstudio/titanium/multiblock/tile/MachineControllerTile.java:
```diff
@@ -0,0 +1,80 @@
+    @Nonnull
+    @Override
+    public T getSelf() {
+        return null;
```
Nonnull, but null. literally return "this"

SkySom commented on src/main/java/com/hrznstudio/titanium/multiblock/tile/MachineFillerTile.java:
```diff
@@ -0,0 +1,39 @@
+
+    @Override
+    public boolean isFormed() {
+        return false;
```
Not formed? just false?

SkySom commented on src/main/java/com/hrznstudio/titanium/multiblock/tile/MachineFillerTile.java:
```diff
@@ -0,0 +1,39 @@
+    @Nonnull
+    @Override
+    public MachineFillerTile getSelf() {
+        return null;
```
... Just "this"

SkySom commented on src/main/java/com/hrznstudio/titanium/util/RotationUtil.java:
```diff
@@ -0,0 +1,31 @@
+            return Rotation.NONE;
+        }
+        if(origin.getAxis() == Direction.Axis.Y || to.getAxis() == Direction.Axis.Y) {
+            return null;
```
Why null here?

SkySom commented on src/main/java/com/hrznstudio/titanium/util/StaticTemplateUtil.java:
```diff
@@ -0,0 +1,58 @@
+
+/**
+ * Full credits for this class to Immersive Engineering
+ * Source: https://github.com/BluSunrize/ImmersiveEngineering/blob/1.14/src/main/java/blusunrize/immersiveengineering/common/blocks/multiblocks/StaticTemplateManager.java
```
That's not generally how licenses work.

SkySom commented on src/main/java/com/hrznstudio/titanium/util/StaticTemplateUtil.java:
```diff
@@ -0,0 +1,58 @@
+        try {
+            return source.getResourceStream(type, id);
+        } catch (IOException e) {
+            throw new RuntimeException(e);
```
Can't you try and avoid crashing the game?

SkySom commented on src/main/java/com/hrznstudio/titanium/util/TitaniumRegistries.java:
```diff
@@ -0,0 +1,22 @@
+public class TitaniumRegistries {
+
+    public static void init(){
+        new RegistryBuilder<IMultiblockHandler>()
```
Why not use the event?

SkySom commented on src/main/java/com/hrznstudio/titanium/util/TitaniumRegistries.java:
```diff
@@ -0,0 +1,22 @@
+                .create();
+    }
+
+    public static IForgeRegistry<IMultiblockHandler> multiblock_registry = RegistryManager.ACTIVE.getRegistry(IMultiblockHandler.class);
```
either needs to be static final and UPPER CASE, or camel case.

SkySom commented on src/main/java/com/hrznstudio/titanium/util/StaticTemplateUtil.java:
```diff
@@ -0,0 +1,58 @@
+        Optional<InputStream> optStream = getModResource(ResourcePackType.SERVER_DATA,
+                new ResourceLocation(id.getNamespace(), id.getPath() + ".nbt"));
+        if (!optStream.isPresent()) {
+            throw new RuntimeException("Mod resource not found: " + id);
```
So much game crashing

Lanse505 committed:
Killed the old API and Refactored

Lanse505 committed:
switched from crash to logging error

Lanse505 committed:
Cleaned-up JavaDocs

Lanse505 committed:
Added proper attribution to any and all classes (I Think)

Lanse505 committed:
Hopefully fixes some of Sky's comments

SkySom commented on src/main/java/com/hrznstudio/titanium/api/multiblock/IMultiblock.java:
```diff
@@ -1,31 +1,63 @@
-/*
```
You removed the license?

Lanse505 replied:
I removed the class and renamed the new one.

Lanse505 committed:
killed old code

Lanse505 committed:
Die code die!

Lanse505 committed:
Cleaned-up Attributions

Lanse505 committed:
Merged RotationUtil -> FacingUtil

Lanse505 committed:
Added License Notifiers

SkySom commented on src/main/java/com/hrznstudio/titanium/Titanium.java:
```diff
@@ -38,6 +38,7 @@
 import com.hrznstudio.titanium.reward.RewardSyncMessage;
 import com.hrznstudio.titanium.reward.storage.RewardWorldStorage;
 import com.hrznstudio.titanium.util.SidedHandler;
+import com.hrznstudio.titanium.util.TitaniumRegistries;
```
Is this actually used.

Lanse505 replied:
It was, not currently

SkySom commented on src/main/java/com/hrznstudio/titanium/api/multiblock/MultiblockTemplate.java:
```diff
@@ -0,0 +1,167 @@
+
+    @Nullable
+    private Template getTemplate() {
+        if (template == null) {
```
You'll probably want to add some sorta other check here, as if things fail, it'll call to load over and over.

SkySom commented on src/main/java/com/hrznstudio/titanium/util/TitaniumRegistries.java:
```diff
@@ -0,0 +1,12 @@
+
+package com.hrznstudio.titanium.util;
+
+public class TitaniumRegistries {
```
It's empyt, but still here?

Lanse505 replied:
Not sure if this will be used in the end to store registry references.
But that was the plan before.

Lanse505 committed:
Killed empty class

Lanse505 committed:
Hopefully fixed broken JavaDocs

Lanse505 committed:
please be fixed

Lanse505 committed:
Push for Drizz

Lanse505 committed:
small push again

drizzs committed:
MultiblockAPI Tile Handling Controller/Filler

Lanse505 committed:
Fixes Drizz broken shit

drizzs committed:
FormationToolHandling

I hope this works!

SkySom commented on src/main/java/com/hrznstudio/titanium/api/multiblock/IFormationItem.java:
```diff
@@ -0,0 +1,7 @@
+package com.hrznstudio.titanium.api.multiblock;
```
Missing License Header

SkySom commented on src/main/java/com/hrznstudio/titanium/api/multiblock/IFormationItem.java:
```diff
@@ -0,0 +1,7 @@
+
+public interface IFormationItem {
+
+    boolean isConsumable();
```
Why can't these be determined by things already existing on the item?

SkySom commented on src/main/java/com/hrznstudio/titanium/api/multiblock/IFormationItem.java:
```diff
@@ -0,0 +1,7 @@
+public interface IFormationItem {
+
+    boolean isConsumable();
+    int formationCost();
```
Why does the Item determining formation cost?

SkySom commented on src/main/java/com/hrznstudio/titanium/api/multiblock/IFormationTool.java:
```diff
@@ -0,0 +1,10 @@
+package com.hrznstudio.titanium.api.multiblock;
```
Missing License header

SkySom commented on src/main/java/com/hrznstudio/titanium/api/multiblock/IFormationTool.java:
```diff
@@ -0,0 +1,10 @@
+import net.minecraft.util.Direction;
+import net.minecraft.util.math.Vec3d;
+
+public interface IFormationTool {
```
Why Formation Tool vs Formation Item?

SkySom commented on src/main/java/com/hrznstudio/titanium/multiblock/tile/MachineControllerTile.java:
```diff
@@ -0,0 +1,183 @@
+    private boolean hasRedStoneSignal = false;
+    @Save
+    private boolean isFormed = false;
+    @Save
```
Why is this saved?

SkySom commented on src/main/java/com/hrznstudio/titanium/multiblock/tile/MachineControllerTile.java:
```diff
@@ -0,0 +1,183 @@
+    private BlockPos masterPos = BlockPos.ZERO;
+    @Save
+    private BlockPos posInMultiblock = BlockPos.ZERO;
+    @Save
```
What if I wanted multiple RS Blocks?

drizzs committed:
Multiblock Controller/Filler, Working it out

This is the basic idea of what i want to do, if you have better idea's on how to implement this let me know!

drizzs committed:
Same as Last!

drizzs committed:
Added Template to Controller Constructor

drizzs committed:
Handling Template to Filler

Lanse505 committed:
small fixes

Lanse505 committed:
run licenceFormat

Lanse505 committed:
Kills breakStructure and replaceStructureBlock

Lanse505 committed:
Kills redundant class

Lanse505 committed:
Refactor

Lanse505 committed:
reformat code

drizzs committed:
Handled Structure Template Check

Lanse505 committed:
shit crashes, yeay!

Lanse505 committed:
fixed crash

Lanse505 committed:
added test template

Lanse505 committed:
fixed derp

drizzs committed:
Item Activation Fix

Lanse505 committed:
Licencing and final touches of the day.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/InnovativeOnlineIndustries/Titanium?pullRequest=30) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/InnovativeOnlineIndustries/Titanium?pullRequest=30) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: Lanse505<br/>:x: drizzs<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/InnovativeOnlineIndustries/Titanium?pullRequest=30) it.</sub>

Lanse505 commented:
I'm killing this PR, might revisit later or talk to Blu about co-opting their system?
